pkg/agent/manager/cache/cache.go

```diff
@@ -688,11 +690,18 @@ func (c *Cache) getSelectorIndex(s selector) *selectorIndex {
 	index, ok := c.selectors[s]
 	if !ok {
 		index = newSelectorIndex()
-		c.selectors[s] = index
+		c.setSelectorIndex(s, index)
 	}
 	return index
 }
 
+// setSelectorIndex sets the selector index for the selector.
+func (c *Cache) setSelectorIndex(s selector, index *selectorIndex) {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+	c.selectors[s] = index
+}
+
```

Comment:
Regardless of change I think it would be good to add a concurrency unit test to both verify existing/new code and prevent future code from accidentally undoing the safety. Like
```
// make cache object c
...
start := make(chan struct{})
var wg sync.WaitGroup

concurrentCallers := 100
callerAttempts := 100
for i := 0; i < concurrentCallers; i++ {
  wg.Add(1)
  go func() {
    defer wg.Done()
    <-start
    for j := 0; j < callerAttempts; j++ {
       // call ops on c that result in `setSelectorIndex`
    }
  }()
}
// start all callers
close(start)
// don't close test until all executions finish
wg.Wait()
```

If we get concurrency panic now or in future in the test, we have broken concurrency.
If there is no concurrency panic, we're 99% sure concurrency is not broken.

Reply:
Absolutely! This package could for sure use a test that shakes out race conditions. However, the suggested test won't be near as effective as it could be until we resolve #2379. I'd suggest we file a separate issue to write this unit test after #2379 is fixed.

Reply:
I think there may still be value, as #2379 would just make the race testing more explicit and global, whereas adding the proposed unit test should be able to catch things today via panic (instead of cleaner race test failure return).

If #2379 is close to completion then sure we can just wait for that.

In any case, no need to flood this PR with such a discussion :)